### PR TITLE
apiextensions: fix missing storage on CRD delete without previous CR access

### PR DESCRIPTION
**Pull request by sttts**
Create CR storage on demand when needed from CRD finalizer controller.

/cc @nikhita

### CODE REVIEW

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -235,39 +212,29 @@ func (r *crdHandler) removeDeadStorage() {
...
-	return ret.storage
+func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions.CustomResourceDefinition) finalizer.ListerCollectionDeleter {
+	info := r.getServingInfoFor(crd)
+	return info.storage
 }
 
 func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefinition) *crdInfo {
-	storageMap := r.customStorage.Load().(crdStorageMap)
-	ret, ok := storageMap[crd.UID]
+	r.customStorageLock.RLock()
```
Comment (deads2k) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: @liggitt specifically requested an atomic.Value is several spots with the claim that it's read performance was better than grabbing a non-contentious read lock.  I don't know one way or the other, but it was an intentional choice.

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go [hunk 2]
```diff
@@ -160,57 +157,35 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
-		handler := handlers.GetResource(storage, storage, requestScope)
-		handler(w, req)
-		return
+		handlers.GetResource(storage, storage, requestScope)(w, req)
 	case "list":
-		forceWatch := false
-		handler := handlers.ListResource(storage, storage, requestScope, forceWatch, minRequestTimeout)
-		handler(w, req)
-		return
+		handlers.ListResource(storage, storage, requestScope, false, minRequestTimeout)(w, req)
```
Comment (deads2k) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: inline comment these at least?  `, false /*forceWatch*/,`

Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: I am lucky that my code editor shows the forceWatch automatically.

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go [hunk 3]
```diff
@@ -235,39 +212,29 @@ func (r *crdHandler) removeDeadStorage() {
...
 // the given uid, or nil if one does not exist.
-func (r *crdHandler) GetCustomResourceListerCollectionDeleter(uid types.UID) finalizer.ListerCollectionDeleter {
-	storageMap := r.customStorage.Load().(crdStorageMap)
-	ret, ok := storageMap[uid]
-	if !ok {
-		return nil
-	}
-	return ret.storage
+func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions.CustomResourceDefinition) finalizer.ListerCollectionDeleter {
+	info := r.getServingInfoFor(crd)
```
Comment (deads2k) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: meat lgtm

**Comment by sttts** [comment]
Will leave the lgtm to @nikhita. If her test works fine with this, please tag it.

**Comment by deads2k** [comment]
> Will leave the lgtm to @nikhita. If her test works fine with this, please tag it.

will do.  @nikhita bug me on slack for a merge.

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Pull request merged**